### ISSUE TITLE
BatchNorm isn't working correctly

### ISSUE THREAD

**Issue by tom-huntington** [hasktorch/hasktorch#626]
It appears that the batchNormWeight and batchNormBias are being updated to be equal to what I'd expect the runningVar and runningMean to be.

```
BatchNorm {batchNormWeight = IndependentTensor {toDependent = Tensor Float [1] [ 0.9001   ]}, batchNormBias = IndependentTensor {toDependent = Tensor Float [1] [ 3.8708e-2]}, runningMean = Tensor Float [1] [ 3.8708e-2], runningVar = Tensor Float [1] [ 0.9001   ]}
```

It think the `at::batch_norm` api isn't doing what's expected.

I made a repro https://github.com/tom-huntington/batch-norm-repro
```
main = do
    batch_norm <- sample $ BatchNormSpec 1
    print batch_norm
    a <- randIO' [2,1]
    print a
    print $ sumDim (Dim 0) KeepDim Float a
    let b = batchNormForward batch_norm True 0.1 0 a
    print b
    print batch_norm
    print $ sumDim (Dim 0) KeepDim Float b
```

**Comment by junjihashimoto** [comment]
@tom-huntington 
I would appreciate it if you could share the details of the expected behavior.

**Comment by tom-huntington** [comment]
Sorry. I think the `batchNormWeight` and `batchNormBias` are being incorrectly updated.

This is the output of the second `print batch_norm`.
```
BatchNorm {batchNormWeight = IndependentTensor {toDependent = Tensor Float [1] [ 0.9148   ]}, batchNormBias = IndependentTensor {toDependent = Tensor Float [1] [ 4.0909e-2]}, runningMean = Tensor Float [1] [ 4.0909e-2], runningVar = Tensor Float [1] [ 0.9148   ]}
```
Notice that the IndependentTensor's have been updated even though no backward pass has been made. Also `batchNormWeight=runningVar`, `batchNormBias=runningMean`.

Changing non-zero values of the momentum (`0.1`  in the above code) doesn't change this behaviour (although 0 momentum gives the expected behaviour).

I think you will have to look into pytorch's source code to figure this out.

**Comment by dmwit** [comment]
Besides the behavior problem described, there needs to be more `IO`s in types somewhere. It shouldn't be possible to get two different results from `show batch_norm` at two different points in the program.

**Comment by junjihashimoto** [comment]
Yeah, we should remove unsafePerformIO of batch_norm(https://github.com/hasktorch/hasktorch/blob/0269df6b3c7fdfa3f25a8c8e315b6188214c57ca/hasktorch/src/Torch/Functional/Internal.hs#L473). The code is generated by codegen.

**Comment by junjihashimoto** [comment]
The code will be like this.
```haskell
{-#LANGUAGE FlexibleContexts#-}
{-#LANGUAGE RecordWildCards#-}
{-#LANGUAGE DeriveGeneric#-}


import System.IO.Unsafe
import Torch.Internal.Cast
import qualified Torch.Internal.Managed.Native as ATen
import GHC.Generics

data BatchNormSpec = BatchNormSpec
  { numFeatures :: Int
  }
  deriving (Show, Eq)

data BatchNorm = BatchNorm
  { batchNormWeight :: Parameter,
    batchNormBias :: Parameter,
    runningMean :: Tensor,
    runningVar :: Tensor
  }
  deriving (Show, Generic)

batchNormIO weight bias running_mean running_var training momentum eps input =
  cast9
      ATen.batch_norm_tttttbddb
      input
      weight
      bias
      running_mean
      running_var
      training
      momentum
      eps
      True

batchNormForwardIO :: BatchNorm -> Bool -> Double -> Double -> Tensor -> IO Tensor
batchNormForwardIO BatchNorm {..} train momentum eps input =
  batchNormIO
    (toDependent batchNormWeight)
    (toDependent batchNormBias)
    runningMean
    runningVar
    train
    momentum
    eps
    input
```

**Comment by junjihashimoto** [comment]
@dmwit 
How about this PR?
https://github.com/hasktorch/hasktorch/pull/649

**Comment by dmwit** [comment]
Hm. Will this still modify one of the input `Tensor`s? If so, I guess I'd find that pretty surprising -- I think that mostly doesn't happen elsewhere in the library. They probably need to be in an `IORef`, or copied, or something.

I put together an implementation [here](https://github.com/dmwit/nurse-sveta/blob/reboot/src/Dr/Mario/Torch.hs#L110-L196) that may be of interest to some people, though admittedly it is not via bindings to the `libtorch` code. But, it does have the advantage of being for the typed API (so the dimensions of the inputs is clarified), and of working cleanly with convolutional nets a la pytorch's `BatchNorm2d` (and `1d`, `3d` via shape polymorphism).

**Comment by junjihashimoto** [comment]
```haskell
newtype MutableTensor = MutableTensor Tensor
newMutableTensor :: Tensor -> IO MutableTensor
newMutableTensor = ...
```
Yeah, IORef is good, and there is also such a newtype .

**Comment by dmwit** [comment]
Ah, nice! Is that in hasktorch somewhere already? I don't see it in [the index](https://hasktorch.github.io/doc-index-M.html); or perhaps you're proposing adding it, in which case I think that's a great idea.

**Comment by junjihashimoto** [comment]
@dmwit 
Hi, sorry for taking so long. It was just a proposal.
The above MutableTensor is merged into mater branch.
https://github.com/hasktorch/hasktorch/pull/649

**Comment by tom-huntington** [comment]
@arnemileswinter just use pytorch. The way you'll learn deep learning is by hacking around on existing real world projects, and they are mostly in pytorch

**Comment by junjihashimoto** [comment]
@arnemileswinter Hi, to be honest, I think it works, but I haven't been able to test it.
typed Resnet is being developed in the following branch, however it is not enough yet.
https://github.com/hasktorch/hasktorch/blob/feature/arrow/examples/resnet/Main.hs#L343-L354
